Add support for composite actions

mdelillo opened the pull request:
This is my first attempt at implementing #339. It supports most features I've tested (using [this repo](https://github.com/mdelillo/act-test)) including using inputs and outputs. One missing feature is using the `Default` value of an input that is not provided.

Before going too far I wanted to get feedback on how to implement some functionality. For convenience I put most of the logic into the `runAction` method, but I'm sure there are better ways to organize this code.

The first problem I ran into is that `${{ github.action_path }}` was getting interpolated to an empty string. The quick fix was just to do `strings.ReplaceAll` on the `Run` field.

The next problem was that outputs from the action were not available due to them being associated with the wrong step. Considering [this example](https://docs.github.com/en/actions/creating-actions/creating-a-composite-run-steps-action), the action declares an output called `random-number`. But the value must be retrieved from the `random-id` output of the `random-number-generator` step.

The hacky way I solved this was to add `OutputMappings` to the `RunContext` which is used to map outputs from action steps to outputs of the action itself. The `CurrentStep` of the `RunContext` also needed to be updated to the ID of the step within the action to know whether the value needs to be mapped. When the action is run, all of the outputs are parsed into this map which is later used by `setOutput`. 

Using the above example, the random number will be stored in `rc.StepResults[foo].Outputs[random-number]` instead of `rc.StepResults[foo].Outputs[random-id]`.

cplee commented:
@mdelillo - this is great, thanks for doing this! Can you add a test workflow file for composite actions with this PR?

wojtekmach commented:
I just wanted to say this is extremely helpful, thank you so much for this. Looking at your test project, I also TIL that you can do `uses: ./<path>` (https://github.com/mdelillo/act-test/blob/d2d86736e41881e90c64ab1846fdc9d278377996/.github/workflows/test.yml#L18) which was super useful too.

TimBobkov commented:
Hope this PR will be merge soon. With composite actions `ACT` will be perfect.

mikemonkers commented on pkg/runner/step_context.go:
```diff
@@ -380,6 +380,43 @@ func (sc *StepContext) runAction(actionDir string, actionPath string) common.Exe
+				stepContext := StepContext{
+					RunContext: rcClone,
+					Step:       &stepClone,
+					Env:        stepClone.Env,
```
I think you need to merge the stepClone.Env with the sc.Env, otherwise all the GITHUB ENV and ACT VARs are lost in the composite:

```
					Env:        mergeMaps(sc.Env, stepClone.Env),
```

mikemonkers commented on pkg/runner/command.go:
```diff
@@ -77,8 +77,15 @@ func (rc *RunContext) setEnv(ctx context.Context, kvPairs map[string]string, arg
-	rc.StepResults[rc.CurrentStep].Outputs[kvPairs["name"]] = arg
+	stepID := rc.CurrentStep
+	outputName := kvPairs["name"]
+	if outputMapping, ok := rc.OutputMappings[MappableOutput{StepID: stepID, OutputName: outputName}]; ok {
```
If there is a mismatch/typo between an output mapping step ID and the actual composite step ID we get a exception/panic due to a null memory reference on line 88 below. eg.

```
outputs:
  tag_name:
    description: "tag to use for the docker image"
    value: ${{ steps.GET_BAG.outputs.tag_name }}

runs:
  using: "composite"
  steps:
    - name: Determine tag
      id: get_tag
      shell: bash
      run: echo "::set-output name=tag_name::latest" 
```

In my foolish case I mixed up '_' and '-' which was hard to spot!

However, it would be nice to have a more informative error than a panic.

mikemonkers commented:
I have built and run it on OSX. Still exercising it for my purposes but some workflow test scenarios would be good too.

JustinGrote commented:
Didn't work for me doing a fairly simple composite test. Maybe it's because I'm on windows and it looks like the slashes get generated backwards but the extraction never happens to the `/actions` folder

## Workflow
```yaml
# yaml-language-server: $schema=https://json.schemastore.org/github-workflow.json
name: Test
on: workflow_dispatch
jobs:
  test:
    name: Composite Test
    runs-on: ubuntu-18.04
    steps:
      - uses: Amadevus/pwsh-script@v2
        name: pwshscript
        id: pwshscript
        with:
          script: |
            throw 'boom'

```

## Docker Exec after act
```
runner@docker-desktop:/github/workflow$ ls /actions
runner@docker-desktop:/github/workflow$ ls 
envs.txt  event.json  script
runner@docker-desktop:/github/workflow$ cat script
\actions\Amadevus-pwsh-script@v2/action.ps1
```

## Logs
```
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading environment from C:\\Users\\JGrote\\Projects\\Press\\.env"
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading secrets from C:\\Users\\JGrote\\Projects\\Press\\.secrets"
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading workflows from 'C:\\Users\\JGrote\\Projects\\Press\\.github\\workflows'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Reading workflow 'C:\\Users\\JGrote\\Projects\\Press\\.github\\workflows\\press.yml'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Reading workflow 'C:\\Users\\JGrote\\Projects\\Press\\.github\\workflows\\test.yml'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Planning event: workflow_dispatch"
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading slug from git directory 'C:\\Users\\JGrote\\Projects\\Press\\.git'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Found revision: 8ca9cff01bf24f3c6f566b092423f7ff034884de\n"
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading revision from git directory 'C:\\Users\\JGrote\\Projects\\Press\\.git'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Found revision: 8ca9cff01bf24f3c6f566b092423f7ff034884de\n"
time="2021-03-23T13:28:36-07:00" level=debug msg="HEAD points to '8ca9cff01bf24f3c6f566b092423f7ff034884de'"
time="2021-03-23T13:28:36-07:00" level=debug msg="HEAD matches refs/heads/ci"
time="2021-03-23T13:28:36-07:00" level=debug msg="using github ref: refs/heads/ci"
time="2021-03-23T13:28:36-07:00" level=debug msg="context env => map[ACT:true]"
[Test/Composite Test] 🚀  Start image=justingrote/act-pwsh
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading slug from git directory 'C:\\Users\\JGrote\\Projects\\Press\\.git'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Found revision: 8ca9cff01bf24f3c6f566b092423f7ff034884de\n"
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading revision from git directory 'C:\\Users\\JGrote\\Projects\\Press\\.git'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Found revision: 8ca9cff01bf24f3c6f566b092423f7ff034884de\n"
time="2021-03-23T13:28:36-07:00" level=debug msg="HEAD points to '8ca9cff01bf24f3c6f566b092423f7ff034884de'"
time="2021-03-23T13:28:36-07:00" level=debug msg="HEAD matches refs/heads/ci"
time="2021-03-23T13:28:36-07:00" level=debug msg="using github ref: refs/heads/ci"
[Test/Composite Test]   🐳  docker pull justingrote/act-pwsh
time="2021-03-23T13:28:36-07:00" level=debug msg="Image exists? true"
[Test/Composite Test]   🐳  docker create image=justingrote/act-pwsh entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[Test/Composite Test]   🐳  docker run image=justingrote/act-pwsh entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[Test/Composite Test] Starting container: 7f954ef4f681a8885d7611bc62477c5e2f80f04fb0c6d72992c339ca8f226e11
[Test/Composite Test] Started container: 7f954ef4f681a8885d7611bc62477c5e2f80f04fb0c6d72992c339ca8f226e11
time="2021-03-23T13:28:36-07:00" level=debug msg="Writing entry to tarball workflow/event.json len:2"
time="2021-03-23T13:28:36-07:00" level=debug msg="Writing entry to tarball workflow/envs.txt len:0"
time="2021-03-23T13:28:36-07:00" level=debug msg="Writing entry to tarball home/.act len:0"
[Test/Composite Test] Extracting content to '/github/'
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading slug from git directory 'C:\\Users\\JGrote\\Projects\\Press\\.git'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Found revision: 8ca9cff01bf24f3c6f566b092423f7ff034884de\n"
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading revision from git directory 'C:\\Users\\JGrote\\Projects\\Press\\.git'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Found revision: 8ca9cff01bf24f3c6f566b092423f7ff034884de\n"
time="2021-03-23T13:28:36-07:00" level=debug msg="HEAD points to '8ca9cff01bf24f3c6f566b092423f7ff034884de'"
time="2021-03-23T13:28:36-07:00" level=debug msg="HEAD matches refs/heads/ci"
time="2021-03-23T13:28:36-07:00" level=debug msg="using github ref: refs/heads/ci"
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading slug from git directory 'C:\\Users\\JGrote\\Projects\\Press\\.git'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Found revision: 8ca9cff01bf24f3c6f566b092423f7ff034884de\n"
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading revision from git directory 'C:\\Users\\JGrote\\Projects\\Press\\.git'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Found revision: 8ca9cff01bf24f3c6f566b092423f7ff034884de\n"
time="2021-03-23T13:28:36-07:00" level=debug msg="HEAD points to '8ca9cff01bf24f3c6f566b092423f7ff034884de'"
time="2021-03-23T13:28:36-07:00" level=debug msg="HEAD matches refs/heads/ci"
time="2021-03-23T13:28:36-07:00" level=debug msg="using github ref: refs/heads/ci"
time="2021-03-23T13:28:36-07:00" level=debug msg="context env => map[ACT:true]"
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading slug from git directory 'C:\\Users\\JGrote\\Projects\\Press\\.git'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Found revision: 8ca9cff01bf24f3c6f566b092423f7ff034884de\n"
time="2021-03-23T13:28:36-07:00" level=debug msg="Loading revision from git directory 'C:\\Users\\JGrote\\Projects\\Press\\.git'"
time="2021-03-23T13:28:36-07:00" level=debug msg="Found revision: 8ca9cff01bf24f3c6f566b092423f7ff034884de\n"
time="2021-03-23T13:28:36-07:00" level=debug msg="HEAD points to '8ca9cff01bf24f3c6f566b092423f7ff034884de'"
time="2021-03-23T13:28:36-07:00" level=debug msg="HEAD matches refs/heads/ci"
time="2021-03-23T13:28:36-07:00" level=debug msg="using github ref: refs/heads/ci"
time="2021-03-23T13:28:36-07:00" level=debug msg="context env => map[ACT:true]"
time="2021-03-23T13:28:36-07:00" level=debug msg="context env => map[ACT:true CI:true GITHUB_ACTION:pwshscript GITHUB_ACTIONS:true GITHUB_ACTOR:nektos/act GITHUB_API_URL:https://api.github.com GITHUB_ENV:/github/workflow/envs.txt GITHUB_EVENT_NAME:workflow_dispatch GITHUB_EVENT_PATH:/github/workflow/event.json GITHUB_GRAPHQL_URL:https://api.github.com/graphql GITHUB_REF:refs/heads/ci GITHUB_REPOSITORY:JustinGrote/Press GITHUB_RUN_ID:1 GITHUB_RUN_NUMBER:1 GITHUB_SERVER_URL:https://github.com GITHUB_SHA:8ca9cff01bf24f3c6f566b092423f7ff034884de GITHUB_TOKEN: GITHUB_WORKFLOW:Test GITHUB_WORKSPACE:/github/workspace HOME:/github/home INPUT_SCRIPT:throw 'boom'\n ImageOS:ubuntu18.04]"
time="2021-03-23T13:28:36-07:00" level=debug msg="setupEnv: map[ACT:true CI:true GITHUB_ACTION:pwshscript GITHUB_ACTIONS:true GITHUB_ACTOR:nektos/act GITHUB_API_URL:https://api.github.com GITHUB_ENV:/github/workflow/envs.txt GITHUB_EVENT_NAME:workflow_dispatch GITHUB_EVENT_PATH:/github/workflow/event.json GITHUB_GRAPHQL_URL:https://api.github.com/graphql GITHUB_REF:refs/heads/ci GITHUB_REPOSITORY:JustinGrote/Press GITHUB_RUN_ID:1 GITHUB_RUN_NUMBER:1 GITHUB_SERVER_URL:https://github.com GITHUB_SHA:8ca9cff01bf24f3c6f566b092423f7ff034884de GITHUB_TOKEN: GITHUB_WORKFLOW:Test GITHUB_WORKSPACE:/github/workspace HOME:/github/home INPUT_SCRIPT:throw 'boom'\n ImageOS:ubuntu18.04]"
[Test/Composite Test] ⭐  Run pwshscript
[Test/Composite Test]   ☁  git clone 'https://github.com/Amadevus/pwsh-script' # ref=v2
[Test/Composite Test]   cloning https://github.com/Amadevus/pwsh-script to act/Amadevus-pwsh-script@v2
[Test/Composite Test] Cloned https://github.com/Amadevus/pwsh-script to act/Amadevus-pwsh-script@v2
[Test/Composite Test] Checked out v2
time="2021-03-23T13:28:37-07:00" level=debug msg="Read action &{PowerShell script Amadevus Run PowerShell scripts within a hydrated context, with workflow-command cmdlets; inspired by actions/github-script.\n map[script:{PowerShell script to execute in Actions-hydrated context true }] map[error:{Exception details, if any was thrown during script execution. ${{ steps.script.outputs.error }}} result:{Return value of script execution ${{ steps.script.outputs.result }}}] {composite map[]   [] [] [{script    ${{ github.action_path }}/action.ps1  pwsh map[PWSH_SCRIPT_ACTION_GITHUB:${{ toJson(github) }} PWSH_SCRIPT_ACTION_JOB:${{ toJson(job) }} PWSH_SCRIPT_ACTION_MATRIX:${{ toJson(matrix) }} PWSH_SCRIPT_ACTION_RUNNER:${{ toJson(runner) }} PWSH_SCRIPT_ACTION_STRATEGY:${{ toJson(strategy) }} PWSH_SCRIPT_ACTION_TEXT:${{ inputs.script }} TEMP:${{ runner.temp }}] map[] false 0}]} {blue terminal}} from 'act\\Amadevus-pwsh-script@v2\\action.yml'"
time="2021-03-23T13:28:37-07:00" level=debug msg="About to run action &{PowerShell script Amadevus Run PowerShell scripts within a hydrated context, with workflow-command cmdlets; inspired by actions/github-script.\n map[script:{PowerShell script to execute in Actions-hydrated context true }] map[error:{Exception details, if any was thrown during script execution. ${{ steps.script.outputs.error }}} result:{Return value of script execution ${{ steps.script.outputs.result }}}] {composite map[]   [] [] [{script    ${{ github.action_path }}/action.ps1  pwsh map[PWSH_SCRIPT_ACTION_GITHUB:${{ toJson(github) }} PWSH_SCRIPT_ACTION_JOB:${{ toJson(job) }} PWSH_SCRIPT_ACTION_MATRIX:${{ toJson(matrix) }} PWSH_SCRIPT_ACTION_RUNNER:${{ toJson(runner) }} PWSH_SCRIPT_ACTION_STRATEGY:${{ toJson(strategy) }} PWSH_SCRIPT_ACTION_TEXT:${{ inputs.script }} TEMP:${{ runner.temp }}] map[] false 0}]} {blue terminal}}"
time="2021-03-23T13:28:37-07:00" level=debug msg="type=3 actionDir=act/Amadevus-pwsh-script@v2 actionPath= Workdir=C:\\Users\\JGrote\\Projects\\Press ActionCacheDir=act actionName=Amadevus-pwsh-script@v2 containerActionDir=/actions"
time="2021-03-23T13:28:37-07:00" level=debug msg="Wrote command '\\actions\\Amadevus-pwsh-script@v2/action.ps1' to 'workflow/script'"
time="2021-03-23T13:28:37-07:00" level=debug msg="Writing entry to tarball workflow/script len:43"
[Test/Composite Test] Extracting content to '/github/'
[Test/Composite Test] Exec command '[pwsh -command "& '/github/workflow/script'"]'
| & '/github/workflow/script'
[Test/Composite Test]   ✅  Success - pwshscript
```

mdelillo commented:
Hi all, thanks for the feedback! I've added a sample workflow and incorporated the changes requested by @mikemonkers.

It still doesn't support default values for inputs in the composite actions, but I'll look into that issue soon.

cplee commented:
@mdelillo - sorry for the confusion... my bad for not being more clear. I was hoping you'd add the test workflow to the [testdata](https://github.com/nektos/act/tree/master/pkg/runner/testdata) directory so that it runs as a part of all the other tests.

cplee reviewed (changes requested):
@mdelillo - sorry for the confusion... my bad for not being more clear. I was hoping you'd add the test workflow to the testdata directory so that it runs as a part of all the other tests.

mikemonkers commented:
@mdelillo I've created a Pull request into your branch with some changes I had made on top of yours including the testdata tests. Hope this helps.

Here's the PR: https://github.com/mdelillo/act/pull/1

mikemonkers commented:
@JustinGrote I think you'll be frustrate with this response but composite actions don't seem support 'uses' (or 'if') in the real world GitHub action runners.

So your attempts to run a powershell action in a composite action:
```
      - uses: Amadevus/pwsh-script@v2
```
... will be thwarted, much as many of my hopes were. :( A change I made in the separate PR to the source of this PR (https://github.com/mdelillo/act/pull/1) adds an corresponding error out to act for this.

catthehacker commented:
@mdelillo Seems like last thing to do is fix conflict in `pkg/model/action.go`.

agentschmitt commented:
@catthehacker The latest release seems not to include this pr. Are there any plans to release this soon? :)

catthehacker commented:
@agentschmitt that's not up to me to decide but you can download artefact including this PR from checks.
![image](https://user-images.githubusercontent.com/31106839/114064763-3d8d7b80-989a-11eb-984e-57a3285c2c46.png)
